packages/context/test/unit/context-observer.unit.ts

```diff
@@ -0,0 +1,319 @@
+      nonMatchingObserver = {
+        filter: binding => false,
+        observe: (event, binding) => {},
+      };
+    }
+  });
+
+  describe('event notification', () => {
+    const events: string[] = [];
+    let nonMatchingObserverCalled = false;
+
+    beforeEach(givenObservers);
+
+    it('emits one bind event to matching observers', async () => {
+      ctx.bind('foo').to('foo-value');
+      await ctx.waitUntilObserversNotified();
+      expect(events).to.eql([
+        'SYNC:foo:foo-value:bind',
+        'ASYNC:foo:foo-value:bind',
+      ]);
+      expect(nonMatchingObserverCalled).to.be.false();
+    });
+
+    it('emits multiple bind events to matching observers', async () => {
+      ctx.bind('foo').to('foo-value');
+      ctx.bind('xyz').to('xyz-value');
+      await ctx.waitUntilObserversNotified();
+      expect(events).to.eql([
+        'SYNC:foo:foo-value:bind',
+        'ASYNC:foo:foo-value:bind',
+        'SYNC:xyz:xyz-value:bind',
+        'ASYNC:xyz:xyz-value:bind',
+      ]);
+    });
+
+    it('emits unbind event to matching observers', async () => {
+      ctx.bind('foo').to('foo-value');
+      await ctx.waitUntilObserversNotified();
+      ctx.unbind('foo');
+      await ctx.waitUntilObserversNotified();
+      expect(events).to.eql([
+        'SYNC:foo:foo-value:bind',
+        'ASYNC:foo:foo-value:bind',
+        'SYNC:foo:foo-value:unbind',
+        'ASYNC:foo:foo-value:unbind',
+      ]);
+      expect(nonMatchingObserverCalled).to.be.false();
+    });
+
+    it('emits bind/unbind events for rebind to matching observers', async () => {
+      ctx.bind('foo').to('foo-value');
+      await ctx.waitUntilObserversNotified();
+      ctx.bind('foo').to('new-foo-value');
+      await ctx.waitUntilObserversNotified();
+      expect(events).to.eql([
+        'SYNC:foo:foo-value:bind',
+        'ASYNC:foo:foo-value:bind',
+        'SYNC:foo:foo-value:unbind',
+        'SYNC:foo:new-foo-value:bind',
+        'ASYNC:foo:foo-value:unbind',
+        'ASYNC:foo:new-foo-value:bind',
+      ]);
+      expect(nonMatchingObserverCalled).to.be.false();
+    });
+
+    it('does not trigger observers if affected binding is the same', async () => {
+      const binding = ctx.bind('foo').to('foo-value');
+      await ctx.waitUntilObserversNotified();
+      expect(events).to.eql([
+        'SYNC:foo:foo-value:bind',
+        'ASYNC:foo:foo-value:bind',
+      ]);
+      ctx.add(binding);
+      await ctx.waitUntilObserversNotified();
+      expect(events).to.eql([
+        'SYNC:foo:foo-value:bind',
+        'ASYNC:foo:foo-value:bind',
+      ]);
+    });
+
+    it('reports error if an observer fails', () => {
+      ctx.bind('bar').to('bar-value');
+      return expect(ctx.waitUntilObserversNotified()).to.be.rejectedWith(
+        'something wrong',
+      );
+    });
+
+    it('does not trigger observers registered after an event', async () => {
+      ctx.bind('foo').to('foo-value');
+      process.nextTick(() => {
```

Comment:
Can you use promise micro-queue instead of next tick?

```suggestion
await Promise.resolve();
```

If not, then please use a promisified version of nextTick.

```ts
// at the top
const nextTickAsync = promisify(process.nextTick);

// here
await nextTickAsync();
```